cupy/core/_reduction.pyx

```diff
@@ -407,29 +896,104 @@ cdef class _AbstractReductionKernel:
             best.user_attrs['block_stride'],
             best.params['out_block_num'])
 
+    def _get_cub_optimized_params(
```

Comment:
@asi1024 Continuing our discussion on the optuna optimization from [here](https://github.com/cupy/cupy/pull/2731#discussion_r435337564), could you take a look at my implementation here? For CUB-based kernels I currently allow two parameters to be tuned: `block_size` and `items_per_thread`. On my local tests I do see Optuna could find better parameters, and I will post the benchmark later.

Reply:
The optimization looks good, but can they be merged somehow?
Maybe, it would be good to have the `use_cub` as a hyperparameter itself?

Reply:
> The optimization looks good, but can they be merged somehow?

I tried, but I realized the number of lines was not reduced, mainly because the bulk of the two `_get*optimized_params` functions, such as `target_func`, `suggest_func`, and `optimize_impl`, are different. Would it help if I pull them out to the module level, combine `_get*optimized_params`, and select which to call based on `use_cub`?

> Maybe, it would be good to have the `use_cub` as a hyperparameter itself?

This is interesting. I haven't thought about it. In my benchmark CUB is a clear winner, but for smaller problem sizes it might help (not sure). But, the hyperparameter would have to be `try_use_cub` instead of `use_cub`, because otherwise some extra work needs to be done for switching back and forth between the two implementations. I suggest we figure out how to deal with your factory function idea first, and then revisit this.